### PR TITLE
New interactive component to replace react-interactive [SATURN-1061]

### PR DESCRIPTION
**Pull request by petesantos**
Code to replace react-interactive. The API is currently the same with the same behaviors.
In addition to testing in the app you can try things out in the storybook story.

Additional notes:
1. Some whitespace changes - be sure to turn on the remove whitespace changes option in the PR. 
2. It looks like _.merge was not playing nicely with forwardRef, this is why the ref is added to props outside of the merge. I will take a deeper look at this issue in lodash to see what is happening
3. A future revision will extract the curated list of styles and inject a style tag into the page, making the interactive component more generic. The current list of properties are based on what we are using in the code

### CODE REVIEW

File: src/components/Interactive.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+const allowedHoverVariables = ['backgroundColor', 'color', 'boxShadow', 'opacity', 'textDecoration']
+const pointerTags = ['button', 'area', 'a', 'select']
+const pointerTypes = ['radio', 'checkbox', 'submit', 'button']
+
+const Interactive = forwardRef(({
+  className = '', as, type, role, onClick, onKeyDown, disabled, children, tabIndex, hover = {}, style = {}, ...props
+}, ref) => {
+  const { cursor } = style
+
+  const onClickPointer = !disabled && (onClick ||
```
Comment (ehanna4) [inline src/components/Interactive.js]: nit: (from mob) consider putting this into a utils.cond rather than ternary

File: src/components/Interactive.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    pointerTags.includes(as) ||
+    pointerTypes.includes(type)) ? { cursor: 'pointer' } : {}
+  const computedCursor = cursor ? { cursor } : onClickPointer
+
+  const onClickTabIndex = onClick ? 0 : undefined
+  const computedTabIndex = _.isNumber(tabIndex) ? tabIndex : onClickTabIndex
```
Comment (ehanna4) [inline src/components/Interactive.js]: (from mob) This is allowing tabbing to disabled elements, we should figure out if we want to allow that (current behavior you cannot). Worth considering allowing tabbing to disabled element so keyboard users can see tooltip.

File: src/components/Interactive.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  const onClickRole = onClick && !['input', ...pointerTags].includes(as) ? 'button' : undefined
+  const computedRole = role ? role : onClickRole
+
+  const cssVariables = _.flow(
+    _.toPairs,
+    _.reduce((result, [key, value]) => {
```
Comment (ehanna4) [inline src/components/Interactive.js]: (from mob) nit: consider using a flatmap instead of reduce and finish with frompairs

**Comment by ehanna4** [comment]
(from mob) Focus ring is staying on elements even after being clicked, need to find solution to handle this. (keep focus ring during tabbing but do not want it after mouse clicking)

File: src/components/Interactive.js (hunk 2 above)
Comment (petesantos) [reply src/components/Interactive.js]: After reviewing some a11y guidelines I kept the behavior as it currently is in Terra. The guidelines recommend excluding disabled elements from the tab order. Also, when trying this out on the workflow page, where there can be many disabled nodes keeping those items in the tab order seemed detrimental to the UX. However, we should take care when adding tooltips to disabled nodes as those tooltip messages will not be accessible to anyone not using a pointing device.

File: src/components/Interactive.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      return [[`--app-hover-${key}`, value], [key, `var(--hover-${key}, ${style[key]})`]]
+    }),
+    _.fromPairs
+  )(hover)
+
+  return h(as, {
+    ref,
+    className: `hover-style ${className}`,
+    style: { ...style, ...cssVariables, cursor: computedCursor, outline },
+    onKeyDown: e => !disabled && onKeyDown ? onKeyDown(e) : e.key === 'Enter' && onClick && onClick(e),
```
Comment (petesantos) [inline src/components/Interactive.js]: Allow the browser to handle the disabled state - remove this check

File: src/components/Interactive.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      return [[`--app-hover-${key}`, value], [key, `var(--hover-${key}, ${style[key]})`]]
+    }),
+    _.fromPairs
+  )(hover)
+
+  return h(as, {
+    ref,
+    className: `hover-style ${className}`,
+    style: { ...style, ...cssVariables, cursor: computedCursor, outline },
+    onKeyDown: onKeyDown || (e => e.key === 'Enter' && onClick && onClick(e)),
```
Comment (panentheos) [inline src/components/Interactive.js]: One more issue I just found: This will call the `onClick` handler on this component, but will not bubble up to higher up components the way an actual click would. Some components (popup menus) are relying on that behavior. We might need to fire an actual browser event here.

File: src/_stories/Interactive.stories.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
+
```
Comment (zarsky-broad) [inline src/_stories/Interactive.stories.js]:
```suggestion
```

File: src/components/Interactive.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import { h } from 'react-hyperscript-helpers'
+import * as Utils from 'src/libs/utils'
+
+
+const allowedHoverVariables = ['backgroundColor', 'color', 'boxShadow', 'opacity', 'textDecoration']
+const pointerTags = ['button', 'area', 'a', 'select']
+const pointerTypes = ['radio', 'checkbox', 'submit', 'button']
+
+const Interactive = forwardRef(({
+  className = '', as, type, role, onClick, onKeyDown, onMouseDown, onBlur, disabled, children, tabIndex, hover = {}, style = {}, ...props
```
Comment (zarsky-broad) [inline src/components/Interactive.js:12]:
```suggestion
  className = '', as, type, role, onClick, onKeyDown, onMouseDown, onBlur, disabled, children, tabIndex, hover = {}, style: { cursor } = {}, ...props
```

File: src/components/Interactive.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+}, ref) => {
+  const [outline, setOutline] = useState(undefined)
+  const { cursor } = style
```
Comment (zarsky-broad) [inline src/components/Interactive.js:15]:
```suggestion
```

File: src/components/NpsSurvey.js [hunk 9]
```diff
@@ -137,18 +137,17 @@ export const NpsSurvey = Utils.connectAtom(authStore, 'authState')(class NpsSurv
           }, 'Submit')
         ])
       ]),
-      h(Clickable, {
-        as: icon('times-circle', {
+      h(Clickable, { onClick: goAway(false) }, [
+        icon('times-circle', {
```
Comment (zarsky-broad) [inline src/components/NpsSurvey.js:141]: Why this change?

Comment (panentheos) [reply src/components/NpsSurvey.js:141]: The old library would accept an element instead of a component for `as`, but we don't.

Comment (zarsky-broad) [reply src/components/NpsSurvey.js:141]: Indeed, I meant why _that_ change

File: src/components/Interactive.js (hunk 7 above)
Comment (petesantos) [reply src/components/Interactive.js:12]: Unfortunately this breaks as I set an empty default style in case style is not passed in. Setting a default and destructuring on the same line does is not working

File: src/components/NpsSurvey.js (hunk 9 above)
Comment (petesantos) [reply src/components/NpsSurvey.js:141]: I would like to discuss this some more, and am planning on making some more revisions (inject styles, handle focus ring better, etc..)

Any objections to leaving it as is for the initial version?

File: src/components/Interactive.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      return [[`--app-hover-${key}`, value], [key, `var(--hover-${key}, ${style[key]})`]]
+    }),
+    _.fromPairs
+  )(hover)
+
+  return h(as, {
+    ref,
+    className: `hover-style ${className}`,
+    style: { ...style, ...cssVariables, cursor: computedCursor, outline },
+    onKeyDown: onKeyDown || (e => e.key === 'Enter' && _.forEach(fn => fn(), [() => e.stopPropagation(), () => e.target.click()])),
```
Comment (petesantos) [inline src/components/Interactive.js]: @panentheos I believe this resolves the issue with the click event bubbling.
Two questions:
1. This does mean anyone passing in an onKeyDown handler will have to add a stopPropagation on "enter" keypresses to their keydown handler if it is in a nested interactive AND they do not want a click bubbling. If they do not add the stopPropagation when the user hits "enter" the keydown event will bubble up to the parent Interactive, triggering a click (if the parent interactive does not have a keydown handler). This is a specific case and it could be something that anyone consuming interactive will need to be aware of. I want to call it out to make sure there are no issues with this. There are 3 keyDown handlers in the code base on searchInput, confirmedSearchInput and in NewUserModal.

2. Thoughts on using the forEach to inline a few function calls dependent on the same condition rather than creating a function with an explicit conditional - is this clear in the code, simpler, etc...

Comment (panentheos) [reply src/components/Interactive.js]: 1. I think that's fine, having nested Interactives is rare enough, never mind with custom key handling. We can always tweak the behavior if we run into it and it makes things awkward.

2. Since we're doing imperative things here, I think just a multiline function with an `if` would be clearer.

File: src/components/NpsSurvey.js (hunk 9 above)
Comment (panentheos) [reply src/components/NpsSurvey.js:141]: Accepting elements would require special detection and handling for that case, which increases the API surface area and the complexity of the component. I think this is a good simplification.

**Review by panentheos — APPROVED** [review]
Looks great, thanks for all your work on this!
